File: dotnet/src/Collate.cs
// Approximation of String.prototype.localeCompare (ICU root collation).
//
// Hashes such as computeSkillFolderHash sort file paths with localeCompare,
// whose order differs from ordinal order (case-insensitive at the primary level,
// punctuation before digits before letters, lowercase before uppercase as a
// tie-break). Matching it keeps computedHash values identical to the TS CLI's.

namespace Skills;

internal static class Collate
{
    /// Primary weight for ASCII punctuation/symbols in CLDR root order.
    private const string PunctOrder = "_-,;:!?.'\"()[]{}@*/\\&#%`^+<=>|~$";

    private static int? Primary(int c)
    {
        if ((c < 0x20 && c is not ('\t' or '\n' or 0x0b or 0x0c or '\r')) || (c >= 0x7f && c < 0xa0)) return null;
        switch (c)
        {
            case '\t': return 1;
            case '\n': return 2;
            case 0x0b: return 3;
            case 0x0c: return 4;
            case '\r': return 5;
            case ' ': return 6;
        }
        if (c < 0x80)
        {
            var i = PunctOrder.IndexOf((char)c);
            if (i >= 0) return 100 + i;
            if (c is >= '0' and <= '9') return 200 + (c - '0');
            if (c is >= 'a' and <= 'z') return 300 + (c - 'a');
            if (c is >= 'A' and <= 'Z') return 300 + (c - 'A');
        }
        if (c is >= 0xD800 and <= 0xDFFF) return 1000 + c; // lone surrogate
        var lower = CodePoints(char.ConvertFromUtf32(c).ToLowerInvariant()).ToList();
        return 1000 + (lower.Count == 1 ? lower[0] : c);
    }

    private static IEnumerable<int> CodePoints(string s)
    {
        for (var i = 0; i < s.Length; i++)
        {
            if (char.IsHighSurrogate(s[i]) && i + 1 < s.Length && char.IsLowSurrogate(s[i + 1]))
            {
                yield return char.ConvertToUtf32(s[i], s[i + 1]);
                i++;
            }
            else
            {
                yield return s[i];
            }
        }
    }

    private static int Tertiary(int c)
    {
        if (c is >= 0xD800 and <= 0xDFFF) return 0;
        var s = char.ConvertFromUtf32(c);
        return s != s.ToLowerInvariant() ? 1 : 0;
    }

    private static int CompareLists(List<int> a, List<int> b)
    {
        for (var i = 0; i < Math.Min(a.Count, b.Count); i++)
        {
            var c = a[i].CompareTo(b[i]);
            if (c != 0) return c;
        }
        return a.Count.CompareTo(b.Count);
    }

    public static int LocaleCompare(string a, string b)
    {
        var pa = CodePoints(a).Select(Primary).Where(p => p.HasValue).Select(p => p!.Value).ToList();
        var pb = CodePoints(b).Select(Primary).Where(p => p.HasValue).Select(p => p!.Value).ToList();
        var c = CompareLists(pa, pb);
        if (c != 0) return c;
        var ta = CodePoints(a).Where(x => Primary(x).HasValue).Select(Tertiary).ToList();
        var tb = CodePoints(b).Where(x => Primary(x).HasValue).Select(Tertiary).ToList();
        return CompareLists(ta, tb);
    }

    /// Stable sort (JS Array.prototype.sort is stable).
    public static List<T> StableSort<T>(IEnumerable<T> items, Comparison<T> cmp) =>
        items.Select((x, i) => (x, i)).OrderBy(t => t, Comparer<(T x, int i)>.Create((l, r) =>
        {
            var c = cmp(l.x, r.x);
            return c != 0 ? c : l.i.CompareTo(r.i);
        })).Select(t => t.x).ToList();

    /// Default JS sort: UTF-16 code unit order.
    public static int CodeUnit(string a, string b) => string.CompareOrdinal(a, b);
}

File: rust/src/collate.rs
//! Approximation of `String.prototype.localeCompare` (ICU root collation).
//!
//! Hashes such as `computeSkillFolderHash` sort file paths with
//! `localeCompare`, whose order differs from byte order (case-insensitive at
//! the primary level, punctuation before digits before letters, lowercase
//! before uppercase as a tie-break). Matching it keeps the Rust port's
//! `computedHash` values identical to the TypeScript CLI's for typical paths.

use std::cmp::Ordering;

/// Primary weight for ASCII punctuation/symbols in CLDR root order.
const PUNCT_ORDER: &str = "_-,;:!?.'\"()[]{}@*/\\&#%`^+<=>|~$";

fn primary(c: char) -> Option<u32> {
    let code = c as u32;
    // Control characters are completely ignorable.
    if code < 0x20 && !matches!(c, '\t' | '\n' | '\u{0b}' | '\u{0c}' | '\r')
        || (0x7f..0xa0).contains(&code)
    {
        return None;
    }
    match c {
        '\t' => return Some(1),
        '\n' => return Some(2),
        '\u{0b}' => return Some(3),
        '\u{0c}' => return Some(4),
        '\r' => return Some(5),
        ' ' => return Some(6),
        _ => {}
    }
    if let Some(i) = PUNCT_ORDER.find(c) {
        return Some(100 + i as u32);
    }
    if c.is_ascii_digit() {
        return Some(200 + (code - '0' as u32));
    }
    if c.is_ascii_alphabetic() {
        return Some(300 + (c.to_ascii_lowercase() as u32 - 'a' as u32));
    }
    let lower: Vec<char> = c.to_lowercase().collect();
    let base = if lower.len() == 1 {
        lower[0] as u32
    } else {
        code
    };
    Some(1000 + base)
}

/// Tertiary weight: lowercase sorts before uppercase.
fn tertiary(c: char) -> u32 {
    if c.is_uppercase() {
        1
    } else {
        0
    }
}

pub fn locale_compare(a: &str, b: &str) -> Ordering {
    let pa: Vec<u32> = a.chars().filter_map(primary).collect();
    let pb: Vec<u32> = b.chars().filter_map(primary).collect();
    match pa.cmp(&pb) {
        Ordering::Equal => {}
        o => return o,
    }
    let ta: Vec<u32> = a
        .chars()
        .filter(|c| primary(*c).is_some())
        .map(tertiary)
        .collect();
    let tb: Vec<u32> = b
        .chars()
        .filter(|c| primary(*c).is_some())
        .map(tertiary)
        .collect();
    ta.cmp(&tb)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn orders_like_icu() {
        let mut v = vec![
            "SKILL.md",
            "references/a.md",
            "README.md",
            "_x",
            "1.txt",
            "b",
            "B",
            "a-b",
            "a_b",
            "ab",
        ];
        v.sort_by(|a, b| locale_compare(a, b));
        assert_eq!(
            v,
            vec![
                "_x",
                "1.txt",
                "a_b",
                "a-b",
                "ab",
                "b",
                "B",
                "README.md",
                "references/a.md",
                "SKILL.md"
            ]
        );
    }
}
